pkg/kubelet/container/helpers.go

```diff
@@ -103,6 +105,14 @@ func HashContainer(container *v1.Container) uint64 {
 	// Omit nil or empty field when calculating hash value
 	// Please see https://github.com/kubernetes/kubernetes/issues/53644
 	containerJSON, _ := json.Marshal(container)
+	if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling) {
+		// In-Place Pod Vertical Scaling allows mutable Resources field.
+		// Changes to this field may not require container restart depending
+		// on policy. So it is excluded from hash.
```

Comment:
I'm not sure we want to change the container hashing function. We should check whether the change was due to resource changes and the restart policy where we do the hash check for a container restart, rather than change the container hashes in a particular case we want to change behaviour for. I'm worried this could have unexpected knock-on effects.

Reply:
Hash is a way to determine if container restart is needed (e.g. image changed). Until now, we have never considered as a class of container spec changes that don't necessarily require restart. So it makes sense to exclude those fields from hash.

Reply:
Currently, when the feature gate is off, we consider the resource requirements for a container as part of its hash spec.

Feature gates can be turned on or off at runtime; this allows us to disable a misbehaving feature without a component restart.

As this is written, turning this feature gate on or off will _globally_ change how container hashes are calculated within that kubelet. If _any_ container specifies resource requirements, this means that it will be restarted as a result of the feature gate toggle.

This is dangerous and unexpected behaviour, since containers that are not using the feature will be affected. Hence this is a blocker for alpha.

Reply:
I thought that toggling feature gate needs a restart of the affected component in order for it to take effect. Is that not true anymore?

How do you dynamically toggle a feature gate? I want to try this out.

Reply:
+1 to @ehashman This is a breaking change.

We should at least calculate hash including and excluding the resources, if any of the hashes matches the current hash, we should not restart the container. And we should always write the new hash so that we can cleanup this behavior in the future.

Reply:
@Random-Liu @ehashman I just remembered something from early prototype. Please see this:
https://github.com/huawei-cloudnative/kubernetes/blob/vertical-scaling/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L534

Is this what you have in mind? (I think I'd call it a more generic ImmutableFieldsHash instead of HashZeroResources)

Reply:
Could you include the change in a commit? It is a bit hard to review and figure out all the changes by looking at that file.

Reply:
Basically what we need is that:
1) If the old container hash and new container hash are different, also check whether they are the same with the resource fields included.
2) If they are still different, restart;
3) If they are the same, not restart, but make sure new hash excludes the resource fields.

Reply:
@Random-Liu @ehashman Please review the incremental change https://github.com/kubernetes/kubernetes/commit/a745b12136027c136e0d3dcace23db99ccdd0d71 which I belive addresses this issue.

Reply:
@vinaykul the linked commit looks good, thank you for your patience in finding this.

@ehashman @Random-Liu thanks for the tips in finding and resolving this.